keanu-project/src/main/java/io/improbable/keanu/util/ProgressBar.java

```diff
@@ -0,0 +1,146 @@
+package io.improbable.keanu.util;
+
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.commons.io.output.NullOutputStream;
+
+public class ProgressBar {
+
+    public static PrintStream DEFAULT_PRINT_STREAM = System.out;
+
+    /**
+     * Disables all progress bars globally
+     */
+    public static void disable() {
+        ENABLED.set(false);
+    }
+
+    /**
+     * Enables all progress bars globally
+     */
+    public static void enable() {
+        ENABLED.set(true);
+    }
+
+    private static ScheduledExecutorService getDefaultScheduledExecutorService() {
+        return Executors.newScheduledThreadPool(1, r -> {
+            Thread t = Executors.defaultThreadFactory().newThread(r);
+            t.setDaemon(true);
+            return t;
+        });
+    }
+
+    private static final AtomicBoolean ENABLED = new AtomicBoolean(true);
+    private static final long FRAME_PERIOD_MS = 500;
+    private static final ProgressUpdate DEFAULT_UPDATE = new ProgressUpdate();
+    private static final String MIDDLE_MESSAGE = "Keanu";
+    private static final String[] FRAMES = new String[]{
+        "|" + MIDDLE_MESSAGE + "|",
+        "\\" + MIDDLE_MESSAGE + "/",
+        "-" + MIDDLE_MESSAGE + "-",
+        "/" + MIDDLE_MESSAGE + "\\"
+    };
+
+    private final ScheduledExecutorService scheduler;
+    private final PrintStream printStream;
+
+    private final AtomicReference<ProgressUpdate> latestProgressUpdate = new AtomicReference<>();
+    private final AtomicInteger nextFrameIndex = new AtomicInteger(0);
+
+    private final List<Runnable> onFinish = new ArrayList<>();
+
+    public ProgressBar(PrintStream printStream, ScheduledExecutorService scheduler) {
+        this.printStream = printStream;
+        this.scheduler = scheduler;
+        startUpdateThread();
+    }
+
+    public ProgressBar(PrintStream printStream) {
+        this(printStream, getDefaultScheduledExecutorService());
+    }
+
+    public ProgressBar() {
+        this(DEFAULT_PRINT_STREAM, getDefaultScheduledExecutorService());
+    }
+
+    public void progress() {
+        progress(DEFAULT_UPDATE);
+    }
+
+    public void progress(String message, Double progressPercentage) {
+        progress(new ProgressUpdate(message, progressPercentage));
+    }
+
+    public void progress(String message) {
+        progress(new ProgressUpdate(message));
+    }
+
+    public void progress(Double progressPercentage) {
+        progress(new ProgressUpdate(progressPercentage));
+    }
+
+    public void progress(ProgressUpdate progressUpdate) {
+        latestProgressUpdate.set(progressUpdate);
```

Comment:
We're creating a new update object Every time one of these things is called even if the progress bar functionality is disabled - should we do this?

Reply:
It's a new object only because it's accessed by two threads and it was my knee jerk reaction. It needs a new object when enabled each time because it should always at least print the last update on finish. But when it's disabled, yes, definitely don't need to go printing or creating anything. I've updated it to only create when enabled.